src/tactic/linarith.lean

```diff
@@ -895,4 +900,55 @@ add_tactic_doc
   decl_names := [`tactic.interactive.linarith],
   tags       := ["arithmetic", "decision procedure", "finishing"] }
 
+/--
+An extension of `linarith` with some preprocessing to allow it to solve some nonlinear arithmetic
+problems. (Based on Coq's `nlinarith` tactic.) See `linarith` for the available syntax of options,
+which are inherited by `nlinarith`; that is, `nlinarith!` and `nlinarith only [h1, h2]` all work as
+in `linarith`. The preprocessing is as follows:
+
+* For every subterm `a ^ 2` in a hypothesis or the goal, the assumption `0 ≤ a ^ 2` is added to
+  the context.
+* For every pair of hypotheses `0 ≤ a`, `0 ≤ b` in the context, the assumption `0 ≤ a * b` is added
+  to the context (non-recursively).
+-/
+meta def tactic.interactive.nlinarith (red : parse ((tk "!")?))
+  (restr : parse ((tk "only")?)) (hyps : parse pexpr_list?)
+  (cfg : linarith_config := {}) : tactic unit := do
+  ls ← match hyps with
+    | none := if restr.is_some then return [] else local_context
+    | some hyps := do
+      ls ← hyps.mmap i_to_expr,
+      if restr.is_some then return ls else (++ ls) <$> local_context
+    end,
+  (s, ge0) ← (list.mfoldr (λ h ⟨s, l⟩, do
+      t ← infer_type h,
+      s ← find_squares s t,
+      return (s, match t with `(0 ≤ %%a) := h :: l | _ := l end))
```

Comment:
Is there any reason to only handle `(0 ≤ %%a)` and not strict inequalities?

Reply:
I don't know, I was just following the description. I think this is quite limited, for example what about general inequalities, don't we want to know `a - b` is nonnegative? If `linarith` is doing context normalization then this pass should probably come afterward, so that it isn't tripped up by a proof of `a >= 0` for example.

Reply:
@robertylewis any opinion?

Reply:
`linarith` is old code that's been extended a bunch of times by various feature requests. I think it would take a nontrivial refactor to adapt the normalization routines to be used here too. Probably the whole preprocessing could just be rewritten.

This is also irrelevant to handling strict inequalities. I've never used `nlinarith` in Coq, but the description really does sound like `linarith` + epsilon. @PatrickMassot you're asking for `linarith` + 2*epsilon. Shouldn't take much extra work to handle >=, <, and so on, so that's fair enough. But I think it's not worth any effort to add features beyond that.

IMO this should be considered a "cheer if it works, don't complain if it doesn't" tactic. There's an endless list of possible "but what if..." feature requests, each one allowing it to solve one more problem.

Reply:
Indeed. I will remind you (@PatrickMassot) that when `nra` was proposed this was the *very first question* I asked about the tactic: what is the extent of what it should be able to do? If we grow it incrementally it will become a hodgepodge (as `linarith` is already, to some extent). I will add support for `0 < a` and `0 <= a` hypotheses, but that's all. Anything more than that should come with an analysis of why it's actually useful and how it contributes to the overall structure, because nonlinear arithmetic, being undecidable, is never going to have a satisfactory tactic covering everything that fits within its nominal scope.

Reply:
> because nonlinear arithmetic, being undecidable

Maybe we're thinking different things by "nonlinear arithmetic" here. The theory of real closed fields is decidable, which is an extension of what a "nonlinear equivalent of linarith" would do. But I don't expect to see a CAD implementation in Lean 3. I agree with you in spirit!

Reply:
I asked Heather to give some examples that she found useful in Coq.